casper/src/main/scala/coop/rchain/casper/Casper.scala

```diff
@@ -239,5 +254,19 @@ sealed abstract class MultiParentCasperInstances {
           case None => ().pure[F]
         }
       }
+
+      private def updateCheckpoints(b: BlockMessage): F[Unit] =
+        Capture[F].capture {
+          val currCheckpoints = checkpoints.take()
+          val (_, newCheckpoints) = InterpreterUtil.computeBlockCheckpoint(
+            b,
+            genesis,
+            blockDag,
+            storageLocation,
+            storageSize,
+            currCheckpoints
+          )
+          checkpoints.put(newCheckpoints)
```

Comment:
`take` followed by the `put` isn't atomic action. Maybe you can  synchronize on the content of `Capture[F].capture`.

Reply:
Hmm, not sure what you mean here. Could you elaborate some more?